File: apps/ios/OfflineInterpreter/Components.swift
import SwiftUI

/// Press-and-hold microphone button. Holding is deliberate: it makes the utterance boundary
/// explicit, which is both faster and more reliable than waiting for silence detection.
struct TalkButton: View {
    let label: String
    let accent: Color
    let active: Bool
    let level: Float
    let enabled: Bool
    let onPress: () -> Void
    let onRelease: () -> Void

    @Environment(\.colorScheme) private var scheme
    @State private var pressing = false

    var body: some View {
        VStack(spacing: 6) {
            Image(systemName: "mic.fill").font(.system(size: 24, weight: .medium))
            Text(label).font(.headline).lineLimit(1).minimumScaleFactor(0.6)
            Text(active ? L("talk_listening") : L("talk_hold"))
                .font(.caption2)
                .opacity(0.75)
                .lineLimit(1)
        }
        .foregroundColor(enabled ? (active ? .white : accent) : Palette.muted(scheme))
        .frame(maxWidth: .infinity)
        .frame(height: 104)
        .background(
            RoundedRectangle(cornerRadius: 22, style: .continuous)
                .fill(enabled ? (active ? accent : accent.opacity(0.16)) : Palette.muted(scheme).opacity(0.12))
        )
        .overlay(
            RoundedRectangle(cornerRadius: 22, style: .continuous)
                .strokeBorder(
                    active ? Color.white.opacity(0.25 + 0.45 * Double(level)) : accent.opacity(0.35),
                    lineWidth: active ? 1 + 3 * CGFloat(level) : 1
                )
        )
        .animation(.easeOut(duration: 0.12), value: active)
        .animation(.easeOut(duration: 0.09), value: level)
        .contentShape(RoundedRectangle(cornerRadius: 22, style: .continuous))
        .gesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in
                    guard enabled, !pressing else { return }
                    pressing = true
                    onPress()
                }
                .onEnded { _ in
                    guard pressing else { return }
                    pressing = false
                    onRelease()
                }
        )
        .disabled(!enabled)
    }
}

/// Level meter drawn as a row of bars — cheap, and readable in bright light.
struct LevelBars: View {
    let level: Float
    let accent: Color
    var bars = 12

    var body: some View {
        HStack(spacing: 3) {
            ForEach(0..<bars, id: \.self) { i in
                let threshold = Float(i + 1) / Float(bars)
                let on = level >= threshold
                RoundedRectangle(cornerRadius: 2)
                    .fill(on ? accent : accent.opacity(0.2))
                    .frame(width: 4, height: on ? CGFloat(8 + 14 * threshold) : 6)
            }
        }
        .animation(.easeOut(duration: 0.08), value: level)
    }
}

struct SectionLabel: View {
    let text: String
    @Environment(\.colorScheme) private var scheme

    var body: some View {
        Text(text.uppercased())
            .font(.caption)
            .kerning(0.4)
            .foregroundColor(Palette.muted(scheme))
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.leading, 4)
            .padding(.bottom, 6)
    }
}

/// Bottom sheet listing the seven languages.
struct LanguagePicker: View {
    let title: String
    let current: String
    let disabled: String?
    let onPick: (String) -> Void

    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationView {
            List(Lang.all) { lang in
                Button {
                    onPick(lang.code)
                    dismiss()
                } label: {
                    HStack {
                        Text(lang.name)
                        Spacer()
                        if lang.code == current {
                            Image(systemName: "checkmark").foregroundColor(.accentColor)
                        }
                    }
                }
                .disabled(lang.code == disabled)
            }
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(L("picker_close")) { dismiss() }
                }
            }
        }
    }
}

File: apps/android/app/src/main/java/com/offlinetranslator/app/ui/Components.kt
package com.offlinetranslator.app.ui

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.offlinetranslator.app.Lang

/**
 * Press-and-hold microphone button. Holding is deliberate: it makes the utterance boundary
 * explicit, which is both faster and more reliable than waiting for silence detection.
 */
@Composable
fun TalkButton(
    label: String,
    accent: Color,
    active: Boolean,
    level: Float,
    enabled: Boolean,
    modifier: Modifier = Modifier,
    onPress: () -> Unit,
    onRelease: () -> Unit,
) {
    val bg by animateColorAsState(
        when {
            !enabled -> MaterialTheme.colorScheme.surfaceVariant
            active -> accent
            else -> accent.copy(alpha = 0.16f)
        },
        tween(120), label = "talkBg",
    )
    val glow by animateFloatAsState(if (active) level.coerceIn(0f, 1f) else 0f, tween(90), label = "talkGlow")
    val content = when {
        !enabled -> MaterialTheme.colorScheme.onSurfaceVariant
        active -> Color.White
        else -> accent
    }

    Box(
        modifier = modifier
            .height(104.dp)
            .clip(RoundedCornerShape(22.dp))
            .background(bg)
            .border(
                width = (1 + 3 * glow).dp,
                color = if (active) Color.White.copy(alpha = 0.25f + 0.45f * glow) else accent.copy(alpha = 0.35f),
                shape = RoundedCornerShape(22.dp),
            )
            .pointerInput(enabled) {
                if (!enabled) return@pointerInput
                detectTapGestures(onPress = {
                    onPress()
                    tryAwaitRelease()
                    onRelease()
                })
            },
        contentAlignment = Alignment.Center,
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Icon(Icons.Filled.Mic, contentDescription = null, tint = content, modifier = Modifier.size(26.dp))
            Spacer(Modifier.height(6.dp))
            Text(
                label,
                style = MaterialTheme.typography.labelLarge,
                color = content,
                textAlign = TextAlign.Center,
                maxLines = 1,
            )
            Text(
                if (active) "말하는 중" else "길게 눌러 말하기",
                style = MaterialTheme.typography.labelMedium,
                color = content.copy(alpha = 0.75f),
                maxLines = 1,
            )
        }
    }
}

/** Level meter drawn as a row of bars — cheap, and readable in bright light. */
@Composable
fun LevelBars(level: Float, accent: Color, bars: Int = 12, modifier: Modifier = Modifier) {
    Row(modifier, horizontalArrangement = Arrangement.spacedBy(3.dp), verticalAlignment = Alignment.CenterVertically) {
        repeat(bars) { i ->
            val threshold = (i + 1f) / bars
            val on = level >= threshold
            Box(
                Modifier
                    .width(4.dp)
                    .height(if (on) (8 + 14 * threshold).dp else 6.dp)
                    .clip(RoundedCornerShape(2.dp))
                    .background(if (on) accent else accent.copy(alpha = 0.20f))
            )
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun LanguagePickerSheet(
    title: String,
    current: String,
    disabled: String?,
    onPick: (String) -> Unit,
    onDismiss: () -> Unit,
) {
    val sheet = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheet) {
        Column(Modifier.padding(bottom = 12.dp).navigationBarsPadding()) {
            Text(
                title,
                style = MaterialTheme.typography.titleMedium,
                modifier = Modifier.padding(start = 24.dp, end = 24.dp, bottom = 8.dp),
            )
            Lang.ALL.forEach { lang ->
                val isDisabled = lang.code == disabled
                Row(
                    Modifier
                        .fillMaxWidth()
                        .pointerInput(lang.code, isDisabled) {
                            detectTapGestures { if (!isDisabled) onPick(lang.code) }
                        }
                        .padding(horizontal = 24.dp, vertical = 14.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(
                        lang.name,
                        style = MaterialTheme.typography.bodyLarge,
                        color = when {
                            isDisabled -> MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.4f)
                            lang.code == current -> MaterialTheme.colorScheme.primary
                            else -> MaterialTheme.colorScheme.onSurface
                        },
                        modifier = Modifier.weight(1f),
                    )
                    if (lang.code == current) {
                        Icon(Icons.Filled.Check, null, tint = MaterialTheme.colorScheme.primary)
                    }
                }
            }
            Spacer(Modifier.height(8.dp))
        }
    }
}

/** Simple section header used by the settings and setup screens. */
@Composable
fun SectionLabel(text: String, modifier: Modifier = Modifier) {
    Text(
        text.uppercase(),
        style = MaterialTheme.typography.labelMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = modifier.padding(start = 4.dp, bottom = 8.dp, top = 4.dp),
    )
}
